vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,476 @@
+            if !entry.message.is_empty() && entry.message.ends_with('\n') {
+                log_handler(EfiDiagnosticsLog {
+                    debug_level,
+                    ticks: time_stamp,
+                    phase,
+                    message: std::mem::take(&mut accumulated_message)
+                        .trim_end_matches(&['\r', '\n'][..])
+                        .to_string(),
+                });
+                entries_processed += 1;
+                is_accumulating = false;
+            }
+
+            // Update bytes read and move to the next entry
+            bytes_read = bytes_read.checked_add(entry.entry_size).ok_or_else(|| {
+                DiagnosticsError::Overflow(anyhow!(
+                    "bytes_read ({}) + entry_size ({})",
+                    bytes_read,
+                    entry.entry_size
+                ))
+            })?;
+
+            // Advance the buffer to the next entry
+            if entry.entry_size >= buffer_slice.len() {
+                break; // End of buffer
+            } else {
+                buffer_slice = &buffer_slice[entry.entry_size..];
+            }
+        }
+
+        // Process any remaining accumulated message
+        if is_accumulating && !accumulated_message.is_empty() {
+            log_handler(EfiDiagnosticsLog {
+                debug_level,
+                ticks: time_stamp,
+                phase,
+                message: std::mem::take(&mut accumulated_message)
+                    .trim_end_matches(&['\r', '\n'][..])
+                    .to_string(),
+            });
+            entries_processed += 1;
+        }
+
+        // Print summary statistics
+        tracelimit::info_ratelimited!(
+            "EFI Diagnostics: Processed {} entries, Read {} bytes",
+            entries_processed,
+            bytes_read
+        );
+
+        Ok(())
+    }
+}
+
+impl UefiDevice {
+    /// Process the diagnostics buffer and log the entries to tracing
+    pub(crate) fn process_diagnostics(&mut self, gm: GuestMemory) {
+        // Do not proceed if we have already processed before
+        if self.service.diagnostics.did_process {
+            tracelimit::warn_ratelimited!("EFI Diagnostics: Already processed, skipping");
+            return;
+        }
+        self.service.diagnostics.did_process = true;
+
+        // Process diagnostics logs and send each directly to tracing
+        match self.service.diagnostics.process_diagnostics(&gm, |log| {
+            tracing::info!(
+                debug_level = log.debug_level,
+                ticks = log.ticks,
+                phase = log.phase,
+                message = %log.message,
+                "EFI Diagnostics:"
+            );
+        }) {
+            Ok(_) => {}
+            Err(error) => {
+                tracelimit::error_ratelimited!(
+                    error = &error as &dyn std::error::Error,
+                    "EFI Diagnostics: Failed to process diagnostics buffer"
```

Comment:
Yeah don't prefix your messages like this. We already have the implicit target (`firmware_uefi::service::diagnostics`) to understand where the message is coming from.